eva-accession-import-automation/genome_downloader.py

```diff
@@ -0,0 +1,171 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from get_assembly_report_url import *
+import os
+import pandas
+import sys
+import urllib.request
+import re
+import json
+import yaml
+from retry import retry
+from __init__ import logger
+
+
+def build_fasta_from_assembly_report(assembly_report_path, assembly_accession, eutils_api_key, directory_path, clear):
+    fasta_path = directory_path + assembly_accession + '.fa'
+    if clear:
+        clear_file_content(fasta_path)
+    written_contigs = get_written_contigs(fasta_path)
+    for chunk in pandas.read_csv(assembly_report_path, skiprows=get_header_line_index(assembly_report_path),
+                                 dtype=str, sep='\t', chunksize=100):
+        new_contigs_in_chunk = process_chunk(chunk, written_contigs, eutils_api_key, fasta_path)
+        written_contigs.extend(new_contigs_in_chunk)
+
+
+def clear_file_content(path):
+    with open(path, 'w'):
+        pass
+
+
+def get_written_contigs(fasta_path):
+    try:
+        written_contigs = []
+        match = re.compile(r'>(.*?)\s')
+        with open(fasta_path, 'r') as file:
+            for line in file:
+                written_contigs.extend(match.findall(line))
+        return written_contigs
+    except FileNotFoundError:
+        logger.info('FASTA file does not exists, starting from scratch')
+        return []
+
+
+def get_header_line_index(assembly_report_path):
+    try:
+        with open(assembly_report_path) as assembly_report_file_handle:
+            return next(line_index for line_index, line in enumerate(assembly_report_file_handle)
+                        if line.lower().startswith("# sequence-name") and "sequence-role" in line.lower())
+    except StopIteration:
+        raise Exception("Could not locate header row in the assembly report!")
+
+
+def process_chunk(assembly_report_dataframe, written_contigs, eutils_api_key, fasta_path):
+    new_contigs = []
+    for index, row in assembly_report_dataframe.iterrows():
+        genbank_accession = row['GenBank-Accn']
+        refseq_accession = row['RefSeq-Accn']
+        relationship = row['Relationship']
+        accession = genbank_accession
+        if relationship != '=' and genbank_accession == 'na':
+            accession = refseq_accession
+        if written_contigs is not None and (genbank_accession in written_contigs or refseq_accession in written_contigs):
+            logger.info('Accession ' + accession + ' already in the FASTA file, don\'t need to be downloaded')
+            continue
+        new_contig = get_sequence_from_ncbi(accession, fasta_path, eutils_api_key)
+        if new_contig is not None:
+            new_contigs.append(new_contig)
+    return new_contigs
+
+
+@retry(tries=3)
```

Comment:
Sorry about being pedantic here but please use a [backoff and a jitter parameter](https://github.com/EBIvariation/eva-cttv-pipeline/blob/a89d4d41cc46885a1e6066699246e254736b7c61/eva_cttv_pipeline/trait_mapping/utils.py#L7). Some traffic managers have DDoS protections that will permaban source IPs that post too many retries successively.

Reply:
You don't sound pedantic at all! Thanks for you comments, they have been very useful!!